File: routes/book_routes.py
from flask import Blueprint, request, jsonify
from models.book import Book  # Adjust the import path as necessary
from mongoengine.errors import NotUniqueError, ValidationError, DoesNotExist

book_routes = Blueprint('book_routes', __name__)

# Route to create a new book
@book_routes.route('/', methods=['POST'])
def add_book():
    try:
        book_data = request.json
        book = Book(**book_data).save()
        return jsonify(book), 201
    except (NotUniqueError, ValidationError) as e:
        return jsonify({'error': str(e)}), 400

# Route to get all books
@book_routes.route('/', methods=['GET'])
def get_books():
    books = Book.objects()
    return jsonify(books), 200

# Route to get a single book by id
@book_routes.route('/<book_id>', methods=['GET'])
def get_book(book_id):
    try:
        book = Book.objects.get(id=book_id)
        return jsonify(book), 200
    except DoesNotExist:
        return jsonify({'error': 'Book not found'}), 404

# Route to update a book by id
@book_routes.route('/<book_id>', methods=['PUT'])
def update_book(book_id):
    try:
        book_data = request.json
        book = Book.objects.get(id=book_id)
        book.update(**book_data)
        return jsonify(book.reload()), 200
    except (DoesNotExist, ValidationError) as e:
        return jsonify({'error': str(e)}), 400

# Route to delete a book by id
@book_routes.route('/<book_id>', methods=['DELETE'])
def delete_book(book_id):
    try:
        book = Book.objects.get(id=book_id)
        book.delete()
        return jsonify({'message': 'Book deleted successfully'}), 200
    except DoesNotExist:
        return jsonify({'error': 'Book not found'}), 404

File: routes/bookRoutes.js
/** 
 * Book Routes
 * 
 * Defines routes for book-related operations within the application. 
 * This includes endpoints to retrieve all books, fetch a single book by ID, 
 * create a new book, update an existing book, and delete a book. 
 * These routes utilize the BookService to interact with the database, 
 * providing an interface for book management.
 * 
 * Routes:
 * POST / - Create a new book
 * GET / - Get all books
 * GET /:id - Get a single book by its unique ID
 * PUT /:id - Update a book's details by its ID
 * DELETE /:id - Deletes a book by its ID
 * 
 * @version 1.0
 * @author Hasan Siddiqui
*/
const express = require('express');
const router = express.Router();
const path = require('path');

// Universal path to the BookService
const BookService = require(path.join(__dirname, '..', 'services', 'BookService'));

// Route to create a new book
router.post('/', async (req, res) => {
    try {
      const newBook = await BookService.addBook(req.body);
      res.status(201).json(newBook);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
});

// Route to get all books
router.get('/', async (req, res) => {
  try {
    const books = await BookService.getAllBooks();
    res.json(books);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Route to get a single book by id
router.get('/:id', async (req, res) => {
  try {
    const book = await BookService.getBookById(req.params.id);
    if (book) {
      res.json(book);
    } else {
      res.status(404).json({ message: 'Book not found' });
    }
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Route to update a book by id
router.put('/:id', async (req, res) => {
  try {
    const updatedBook = await BookService.updateBook(req.params.id, req.body);
    if (updatedBook) {
      res.json(updatedBook);
    } else {
      res.status(404).json({ message: 'Book not found' });
    }
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
});

// Route to delete a book by id
router.delete('/:id', async (req, res) => {
  try {
    const bookDeleted = await BookService.deleteBook(req.params.id);
    if (bookDeleted) {
      res.json({ message: 'Book deleted successfully' });
    } else {
      res.status(404).json({ message: 'Book not found' });
    }
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

module.exports = router;
